### PR TITLE
#167727635 Build multi-city request route

### PR DESCRIPTION
**Pull request by victor-shagor**
#### What does this PR do?
* allows user to book multi city trips
#### Description of Task to be completed?
* implement a multicity request endpoint
#### How should this be manually tested?
* Clone the repo and `cd` into the project directory
* Run `yarn install` to install the project's dependency
* Run `yarn start` to start the app
* open this route POST /api/v1/request/multicity on postman

#### What are the relevant pivotal tracker stories?
[#167727635](https://www.pivotaltracker.com/story/show/167727635)

#### Any background context you want to add?
* You must be signed in to access this feature
 - You need to pass in the following information to the request body
 ````
* Note:  destination and flightDate must be an array
 {
  origin: string
  destination: []
  flightDate: []
}
````

### CODE REVIEW

File: src/models/Destination.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+      type: DataTypes.STRING,
+      // validate: {
+      //   notEmpty: {
+      //     msg: 'Destination is required.',
+      //   }
+      // }
+    },
+    multiDestination: {
+      type: DataTypes.ARRAY(DataTypes.STRING),
+      // validate: {
```
Comment (gmemmy) [inline src/models/Destination.js]: Hey, great job! Why do you have lines of commented out code?

**Review by gmemmy — CHANGES_REQUESTED** [review]

File: src/models/Destination.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
+export default (sequelize, DataTypes) => {
+  const Destination = sequelize.define('Destination', {
```
Comment (gmemmy) [inline src/models/Destination.js]: Also is the creation of a new schema called *Destination*  necessary? all these can fit into the *Request* model. Then again, *Destination* isn't descriptive enough of what the schema is about if at all it is necessary. From what I can see, the only difference is the addition of a *multiDestination* row, which can easily fit into the Request table. *multiFlightDate* too can still fit into the Request table. My point is, the multi-city request is a request and should be in the request table. Great job!!!!

File: src/models/Destination.js (hunk 1 above)
Comment (victor-shagor) [reply src/models/Destination.js]: will adjust that now, thanks

File: src/models/Destination.js (hunk 2 above)
Comment (victor-shagor) [reply src/models/Destination.js]: that make sense, will try and implement that

File: src/validation/index.js [hunk 3]
```diff
@@ -25,6 +25,19 @@ const allFieldsRequired = (res, message) => {
...
+ * Defines the failed message returned when required fields are missing.
+ * @param {object} res - Response object
+ * @param {string} message - specific error message
+ * @returns {res} - Response object
+ */
+const errorMsg = (res, message) => {
+  res.status(400).send({
+    success: false,
+    message,
+  });
```
Comment (ChidozieGabriel) [inline src/validation/index.js]: This method already exists. Please do use the `clientError` method in the `utils/helperMethod.js` file.

**Review by ChidozieGabriel — CHANGES_REQUESTED** [review]
Good code @victor-shagor  :+1: 
Please consider this review.

File: src/validation/index.js (hunk 3 above)
Comment (victor-shagor) [reply src/validation/index.js]: done thanks

File: src/models/Destination.js (hunk 1 above)
Comment (victor-shagor) [reply src/models/Destination.js]: adjusted, thanks

File: src/models/Destination.js (hunk 2 above)
Comment (victor-shagor) [reply src/models/Destination.js]: done , thanks

File: src/validation/index.js [hunk 4]
```diff
@@ -121,6 +123,25 @@ class Validate {
...
+  /**
+  * @param {object} req - Request object
+  * @param {object} res - Response object
+  * @param {callback} next - The callback that passes the request to the next handler
+  * @returns {object} res - Response object when query is invalid
+  * @memberof Validate
+  */
+  static validateMulticity(req, res, next) {
+    const { destination, flightDate } = req.body;
+    if (!destination || typeof (destination) !== 'object') {
```
Comment (gmemmy) [inline src/validation/index.js]: Is destination an object or an array?

**Review by gmemmy — DISMISSED** [review]

File: src/controllers/RequestController.js [hunk 5]
```diff
@@ -21,7 +21,13 @@ class RequestController {
     try {
       const { id } = req.decoded;
       const { body } = req;
-      const { dataValues } = await Request.create({ ...body, userId: id });
+      const {
+        // eslint-disable-next-line camelcase
+        origin, returnDate, return_trip, destination, flightDate
```
Comment (Daymorelah) [inline src/controllers/RequestController.js]: Change `return_trip` to `returnTrip`. Remember to change it everywhere it occurs. e.g. models, migrations, tests

File: src/controllers/RequestController.js [hunk 6]
```diff
@@ -66,6 +72,42 @@ class RequestController {
       return HelperMethods.serverError(res);
     }
   }
+
+  /**
+  * Book a Trip
```
Comment (Daymorelah) [inline src/controllers/RequestController.js:126]: Bros change this to `BookMulticity`. You see what copy and paste can cause?.

File: src/migrations/03-create-requests.js [hunk 7]
```diff
@@ -40,6 +43,10 @@ module.exports = {
         key: 'id'
       }
     },
+    return_trip: {
+      type: Sequelize.BOOLEAN,
+      allowNull: true,
```
Comment (Daymorelah) [inline src/migrations/03-create-requests.js]: had a default value here which indicates that a request is not a return trip. Thus remove the allowNull attribute.

File: src/validation/index.js [hunk 8]
```diff
@@ -121,6 +123,25 @@ class Validate {
...
+      return HelperMethods.clientError(res, 'destination as to be more than one');
```
Comment (Daymorelah) [inline src/validation/index.js]: What if it is a single trip? Don't you want to say `should be at least one destination`?

**Review by Daymorelah — CHANGES_REQUESTED** [review]

File: src/validation/index.js (hunk 8 above)
Comment (victor-shagor) [reply src/validation/index.js]: This middleware only validates multicity route

File: src/migrations/03-create-requests.js (hunk 7 above)
Comment (victor-shagor) [reply src/migrations/03-create-requests.js]: done..thanks

File: src/controllers/RequestController.js (hunk 5 above)
Comment (victor-shagor) [reply src/controllers/RequestController.js]: done

File: src/controllers/RequestController.js [hunk 9]
```diff
@@ -21,7 +21,12 @@ class RequestController {
     try {
       const { id } = req.decoded;
       const { body } = req;
-      const { dataValues } = await Request.create({ ...body, userId: id });
+      const {
+        origin, returnDate, returnTrip, destination, flightDate
+      } = body;
```
Comment (Daymorelah) [inline src/controllers/RequestController.js]: Lines 22-26 is not necessary. just do: `{ oriign, returnDate, ...other fileds, } = req.body;`

File: src/controllers/RequestController.js [hunk 10]
```diff
@@ -66,6 +71,41 @@ class RequestController {
...
+  * @return {res} res - HTTP Response object
+  * @memberof RequestController
+ */
+  static async bookMulticity(req, res) {
+    try {
+      const { id } = req.decoded;
+      const { body } = req;
+      const {
+        origin, returnDate, returnTrip, destination, flightDate
+      } = body;
```
Comment (Daymorelah) [inline src/controllers/RequestController.js]: Edit lines 86-89 as I explained above.

**Review by Daymorelah — CHANGES_REQUESTED** [review]

File: src/validation/index.js (hunk 8 above)
Comment (Daymorelah) [reply src/validation/index.js]: okay

**Comment by Daymorelah** [comment]
Please update your PR description. How would your PR get tested if you don't explain how to test it in your PR description?

File: src/test/integrationTests/requestController.test.js [hunk 11]
```diff
@@ -74,24 +75,121 @@ describe('Integration tests for the request controller', () => {
...
+        returnTrip: true,
+        userId: '79ddfd3b-5c83-4beb-815e-55b1c95230e1',
+        reason: 'EXPEDITION',
+      });
+    expect(response.status).to.equal(400);
+    expect(response.body).to.have.property('message');
+    expect(response.body).to.have.property('success');
+    expect(response.body.success).to.equal(false);
+    expect(response.body.message).to.equal('destination as to be more than one');
+  });
```
Comment (Daymorelah) [inline src/test/integrationTests/requestController.test.js]: Why are you repeating the test cases?. Lines 117 and 131 are testing for the same thing. If you are testing for different required fields then the description of your `it` block should be `should not book a trip when the "destination" field is missing`

**Review by Daymorelah — CHANGES_REQUESTED** [review]

File: src/test/integrationTests/requestController.test.js (hunk 11 above)
Comment (victor-shagor) [reply src/test/integrationTests/requestController.test.js]: okay, i will edit that now

File: package.json [hunk 12]
```diff
@@ -5,7 +5,7 @@
   "main": "index.js",
   "scripts": {
     "coverage": "nyc report --reporter=text-lcov | coveralls",
-    "test": "NODE_ENV=test yarn undo:migration && NODE_ENV=test yarn migrate && NODE_ENV=test yarn undo:seed && NODE_ENV=test yarn seed && NODE_ENV=test nyc --reporter=html --reporter=text mocha --exit --require @babel/register --require @babel/polyfill src/test/index.js",
+    "test": "NODE_ENV=test yarn undo:migration && NODE_ENV=test yarn migrate && NODE_ENV=test yarn undo:seed && NODE_ENV=test yarn seed && NODE_ENV=test nyc --reporter=html --reporter=text mocha --exit --require @babel/register --require @babel/polyfill src/test/index.js --timeout 3000 --exit",
```
Comment (ChidozieGabriel) [inline package.json]: You have duplicate `--exit`.
Remove one:
```suggestion
    "test": "NODE_ENV=test yarn undo:migration && NODE_ENV=test yarn migrate && NODE_ENV=test yarn undo:seed && NODE_ENV=test yarn seed && NODE_ENV=test nyc --reporter=html --reporter=text mocha --exit --require @babel/register --require @babel/polyfill src/test/index.js --timeout 3000",
```

File: src/test/integrationTests/requestController.test.js [hunk 13]
```diff
@@ -74,24 +75,121 @@ describe('Integration tests for the request controller', () => {
...
       expect(response.body).to.have.property('success');
       expect(response.body.success).to.equal(false);
     });
   });
+  it('should allow a registered user to book a multcity trip', async () => {
+    const response = await chai.request(app).post('/api/v1/request/multicity')
+      .set('x-access-token', token).send({
+        origin: 'Onipan',
+        destination: ['mile12', 'okoko'],
+        flightDate: ['2019-06-27', '2019-06-25'],
```
Comment (ChidozieGabriel) [inline src/test/integrationTests/requestController.test.js]: What happens when user enters inconsistent _destination_ and _flightDate_ properties? An example is given below:
```
{
  ...
  destination: ['mile12', 'okoko'],
  flightDate: ['2019-06-27'],
}
```

Did you consider this in your error catching? If not, please do. A good error message could be _'Destination places and flight dates do not match.'_; something like that.

The code could be added in the `validateMulticity` validation method.
```
static validateMulticity(req, res, next) {
  ...
  // validate destination and flightDate consistency...
}
```

Thanks.

**Review by ChidozieGabriel — CHANGES_REQUESTED** [review]
> Great code :+1: 

@victor-shagor Please consider this review. Thanks.

**Comment by ChidozieGabriel** [comment]
@victor-shagor Please do NOT merge develop into your feature branch when you have merge conflicts. Use the appropriate git workflow:
- git fetch origin
- git rebase origin/develop
- git push --force origin my-new-feature

[Andela Git Workflow](https://github.com/andela/bestpractices/wiki/Git-Workflow#the-workflow)

File: src/test/integrationTests/requestController.test.js (hunk 13 above)
Comment (victor-shagor) [reply src/test/integrationTests/requestController.test.js]: i did consider it thanks

**Pull request merged**